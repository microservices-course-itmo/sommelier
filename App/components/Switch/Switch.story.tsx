import React from 'react'
import { storiesOf } from '@storybook/react-native/dist'
import { withKnobs, boolean, select } from '@storybook/addon-knobs'
import Switch from './switch'
import { BufferView, ThemeDecorator } from '../../config/decorators'

const defaultSwitch = () => (
  <Switch
    type={select(
      'Type',
      {
        something: 'something',
        heart: 'heart',
      },
      'heart'
    )}
    value={boolean('Value', true)}
    disabled={boolean('Disabled', false)}
  />
)

storiesOf('Switch', module)
  .addDecorator(BufferView)
  .addDecorator(ThemeDecorator)
  .addDecorator(withKnobs)
  .add('default', defaultSwitch)
