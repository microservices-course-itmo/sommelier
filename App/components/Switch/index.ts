export { Switch } from './switch'
export { SwitchProps } from './switch.types'
