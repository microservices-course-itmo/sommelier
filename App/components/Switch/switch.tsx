import React, { useEffect, useState } from 'react'
import FilledFavorite from '../../../assets/images/filledFavorite.svg'
import UnfilledFavorite from '../../../assets/images/unfilledFavorite.svg'
import CheckMark from '../../../assets/images/checkMark.svg'
import * as Styled from './switch.styled'
import theme from '../../theme'
import { Props } from './switch.types'

const Switch = ({
  switchType = 'default',
  value = false,
  onChange = (): void => {},
  disabled = false,
}: Props) => {
  const [stateValue, setStateValue] = useState(value)

  useEffect(() => {
    onChange(stateValue)
  }, [stateValue])

  const toggleSwitch = () => {
    if (!disabled) {
      setStateValue(!stateValue)
    }
  }

  return (
    <Styled.SwitchContainer
      onPress={toggleSwitch}
      value={stateValue}
      disabled={disabled}
      testID='sommelier-switch'
    >
      {(() => {
        switch (switchType) {
          case theme.switchTypes.heart: {
            if (stateValue) {
              return <FilledFavorite width='25' height='25' />
            }
            return <UnfilledFavorite width='25' height='25' />
          }
          case theme.switchTypes.default: {
            return (
              <Styled.Toggle value={stateValue}>
                <Styled.Notch value={stateValue}>
                  {stateValue && (
                    <Styled.Mark>
                      <CheckMark />
                    </Styled.Mark>
                  )}
                </Styled.Notch>
              </Styled.Toggle>
            )
          }
          default: {
            return null
          }
        }
      })()}
    </Styled.SwitchContainer>
  )
}

export default Switch
