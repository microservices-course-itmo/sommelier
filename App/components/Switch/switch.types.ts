import { ReactNode } from 'react'
import { GestureResponderEvent } from 'react-native'

export type Props = {
  children?: ReactNode
  type?: 'heart' | 'something'
  value?: boolean
  onChange?: (value: boolean) => void
  onClick?: (value: boolean, event: GestureResponderEvent) => void
  disabled?: boolean
}
