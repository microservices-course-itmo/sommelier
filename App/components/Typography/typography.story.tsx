import React from 'react'
import styled from 'styled-components/native'
import { storiesOf } from '@storybook/react-native'
import { boolean, select } from '@storybook/addon-knobs'
import { H1, H2, H3, H4, H5, H6, Text } from './typography'
import { BufferView, ThemeDecorator } from '../../config/decorators'
import theme from '../../theme/index'

const StyledWrapper = styled.View`
  display: flex;
  flex-direction: column;
`

const FONT_SIZES = Object.keys(theme.fontSizes) as Array<
  keyof typeof theme.fontSizes
>
const COLORS = Object.keys(theme.colors) as Array<keyof typeof theme.colors>
const FONTS = Object.keys(theme.fonts) as Array<keyof typeof theme.fonts>

export const list = () => (
  <StyledWrapper>
    <H1 color='black' bold={false} italic={false} fontFamily='PtSans'>
      H1 value
    </H1>
    <H2 color='black' bold={false} italic={false} fontFamily='PtSans'>
      H2 value
    </H2>
    <H3 color='black' bold={false} italic={false} fontFamily='PtSans'>
      H3 value
    </H3>
    <H4 color='black' bold={false} italic={false} fontFamily='PtSans'>
      H4 value
    </H4>
    <H5 color='black' bold={false} italic={false} fontFamily='PtSans'>
      H5 value
    </H5>
    <H6 color='black' bold={false} italic={false} fontFamily='PtSans'>
      H6 value
    </H6>
    <Text
      color='black'
      bold={false}
      italic={false}
      fontFamily='PtSans'
      size='md'
    >
      Text value
    </Text>
  </StyledWrapper>
)

export const knobs = () => (
  <StyledWrapper>
    <H1
      color={select('color', COLORS, COLORS[1])}
      bold={boolean('bold', false)}
      italic={boolean('font-style: italic', false)}
      fontFamily={select('font-family', FONTS, FONTS[0])}
    >
      H1 value
    </H1>
    <H2
      color={select('color', COLORS, COLORS[1])}
      bold={boolean('bold', false)}
      italic={boolean('font-style: italic', false)}
      fontFamily={select('font-family', FONTS, FONTS[0])}
    >
      H2 value
    </H2>
    <H3
      color={select('color', COLORS, COLORS[1])}
      bold={boolean('bold', false)}
      italic={boolean('font-style: italic', false)}
      fontFamily={select('font-family', FONTS, FONTS[0])}
    >
      H3 value
    </H3>
    <H4
      color={select('color', COLORS, COLORS[1])}
      bold={boolean('bold', false)}
      italic={boolean('font-style: italic', false)}
      fontFamily={select('font-family', FONTS, FONTS[0])}
    >
      H4 value
    </H4>
    <H5
      color={select('color', COLORS, COLORS[1])}
      bold={boolean('bold', false)}
      italic={boolean('font-style: italic', false)}
      fontFamily={select('font-family', FONTS, FONTS[0])}
    >
      H5 value
    </H5>
    <H6
      color={select('color', COLORS, COLORS[1])}
      bold={boolean('bold', false)}
      italic={boolean('font-style: italic', false)}
      fontFamily={select('font-family', FONTS, FONTS[0])}
    >
      H6 value
    </H6>
    <Text
      color={select('color', COLORS, COLORS[1])}
      bold={boolean('bold', false)}
      italic={boolean('font-style: italic', false)}
      fontFamily={select('font-family', FONTS, FONTS[0])}
      size={select('text font-size', FONT_SIZES, FONT_SIZES[0])}
      crossedOut={boolean('cross-out', false)}
    >
      Text value
    </Text>
  </StyledWrapper>
)

storiesOf('Typography', module)
  .addDecorator(BufferView)
  .addDecorator(ThemeDecorator)
  .add('Typography list', list)

storiesOf('Typography', module)
  .addDecorator(BufferView)
  .addDecorator(ThemeDecorator)
  .add('Typography knobs', knobs)
