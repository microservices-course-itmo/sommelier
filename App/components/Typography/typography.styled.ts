import styled from 'styled-components/native'
import {
  TypographyProps,
  StyledHeaderProps,
  StyledTextProps,
} from './typography.types'

const StyledTypography = styled.Text<TypographyProps>`
  color: ${({ color, theme }) =>
    color ? theme.colors[color] : theme.colors.black};
  font-weight: ${({ bold }) => (bold ? 'bold' : 'normal')};
  font-style: ${({ italic }) => (italic ? 'italic' : 'normal')};
  text-decoration: ${({ crossedOut }) =>
    crossedOut ? 'line-through' : 'none'};
`

export const StyledHeader = styled(StyledTypography)<StyledHeaderProps>`
  font-size: ${({ size, theme }) => theme.headerSizes[size]};
`

export const StyledText = styled(StyledTypography)<StyledTextProps>`
  font-size: ${({ size, theme }) => theme.fontSizes[size]};
`
