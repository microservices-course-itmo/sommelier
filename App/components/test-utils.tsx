import React from 'react'
import { render } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components'
import theme from '../theme'

const themeRender = (component: any) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}

export * from '@testing-library/react-native'

export { themeRender }
