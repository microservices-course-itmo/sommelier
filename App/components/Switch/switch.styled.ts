import styled from 'styled-components/native'
import { Props, ToggleProps, NotchProps } from './switch.types'

export const SwitchContainer = styled.TouchableWithoutFeedback<Props>`
  ${({ disabled }) => disabled && `opacity: 0.5;`}
`

export const Toggle = styled.View<ToggleProps>`
  height: 16px;
  width: 34px;
  border-radius: 50px;
  background-color: ${({ value }) =>
    value ? 'rgba(69, 208, 158, 0.5)' : '#ededed'};
`

export const Notch = styled.View<NotchProps>`
  height: 20px;
  width: 20px;
  border-radius: 10px;
  background-color: ${({ value }) => (value ? '#45d09e' : '#e0e0e0')};
  ${({ value }) => (value ? 'right: -3px;' : 'left: -3px;')}
  top: -2px;
  position: absolute;
  z-index: 1;
`

export const Mark = styled.View`
  position: absolute;
  left: 6px;
  top: 6px;
`
