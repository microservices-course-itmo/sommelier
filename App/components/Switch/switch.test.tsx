import React from 'react'
import { themeRender } from '../../config/test-utils'
import Switch from './switch'
import { Text } from '../Typography'

describe('Switch component', () => {
  const mockFn = jest.fn()
  it('Component with only necessary props matches snapshot', () => {
    const tree = themeRender(<Switch />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('Component with all props matches snapshot', () => {
    const tree = themeRender(
      <Switch
        switchType='heart'
        value
        onChange={mockFn}
        onClick={mockFn}
        disabled
      >
        <Text>abc</Text>
      </Switch>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  // it('SwitchType prop assigns correctly', () => {
  //   const elem = themeRender(<Switch switchType='heart' />).toJSON()
  //   fireEvent.press(elem)
  //   expect(mockFn).toHaveBeenCalled()
  // })
})
