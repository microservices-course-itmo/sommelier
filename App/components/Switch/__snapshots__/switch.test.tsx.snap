// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component Component with all props matches snapshot 1`] = `
<SvgMock
  accessible={true}
  focusable={true}
  height="25"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  width="25"
/>
`;

exports[`Switch component Component with only necessary props matches snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#ededed",
        "borderRadius": 50,
        "height": 16,
        "width": 34,
      },
    ]
  }
  value={false}
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e0e0e0",
          "borderRadius": 10,
          "height": 20,
          "left": -3,
          "position": "absolute",
          "top": -2,
          "width": 20,
          "zIndex": 1,
        },
      ]
    }
    value={false}
  />
</View>
`;
